packages/typescript-estree/src/parser.ts

```diff
@@ -347,6 +366,41 @@ function warnAboutTSVersion(): void {
   }
 }
 
+/**
+ * ESLint (and therefore typescript-eslint) is used in both "single run"/one-time contexts,
+ * such as an ESLint CLI invocation, and long-running sessions (such as continuous feedback
+ * on a file in an IDE).
+ *
+ * When typescript-eslint handles TypeScript Program management behind the scenes, this distinction
+ * is important because there is significant overhead to managing the so called Watch Programs
+ * needed for the long-running use-case. We therefore use the following logic to figure out which
+ * of these contexts applies to the current execution.
+ */
+function inferSingleRun(): void {
+  // Allow users to explicitly inform us of their intent to perform a single run (or not) with TSESTREE_SINGLE_RUN
+  if (process.env.TSESTREE_SINGLE_RUN === 'false') {
+    extra.singleRun = false;
+    return;
+  }
+
+  if (
+    process.env.TSESTREE_SINGLE_RUN === 'true' ||
+    // Default to single runs for CI processes. CI=true is set by most CI providers by default.
+    process.env.CI === 'true' ||
+    // This will be true for invocations such as `npx eslint ...` and `./node_modules/.bin/eslint ...`
+    process.argv[1].endsWith(normalize('node_modules/.bin/eslint'))
+  ) {
+    extra.singleRun = true;
+    return;
+  }
+
+  /**
+   * We default to assuming that this run could be part of a long-running session (e.g. in an IDE)
+   * and watch programs will therefore be required
+   */
+  extra.singleRun = false;
```

Comment:
logic LGTM - question about how we want to roll this out.
We might want to allow a week or two of opting in to this feature so that we can have some repos test this.

I was thinking we should add `parserOptions.allowAutomaticSingleRunInference`, which would disable the second `if`.

Eg:

```ts
if (TSESTREE_SINGLE_RUN is false) { return off }
if (TSESTREE_SINGLE_RUN is true) { return on }
if (parserOptions.allowAutomaticSingleRunInference is false) { return off }
if (automatic inference) { return on }
```

That way we can land this this week with the feature disabled, share it on twitter and get a few repos to test, and then turn it on once they've confirmed it's all good.

My reasoning is that it just gives us that safety net so that we can mitigate the chance of another bad rollout like we had with the 3.0 changes.

Reply:
Sure, sounds reasonable